### PR TITLE
feat(controller): add simple blocking HUD

### PR DESCRIPTION
**Pull request by Polostor**
Added initial HUD without actions #22
Added node for minimap #100

### CODE REVIEW

File: crates/controller/src/hud.rs [hunk 1]
```diff
@@ -79,3 +89,30 @@ fn process_events(
...
+    commands.spawn(
+        NodeBundle {
+            style: Style {
+                justify_content: JustifyContent::SpaceBetween,
+                flex_direction: FlexDirection::Row,
+                align_items: AlignItems::FlexEnd,
+                position_type: PositionType::Absolute,
+                size: Size {
+                    width: Val::Percent(100.0),
+                    height: Val::Percent(100.0),
```
Comment (Indy2222) [inline crates/controller/src/hud.rs:105]: Cannot the height be set to the height of the tallest component?

File: crates/controller/src/hud_components.rs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    commands: &mut ChildBuilder,
+) {
+    commands.spawn((side_node_bundle(HUD_COLOR), HudVisibleNode));
+}
+
+fn side_node_bundle(color: Color) -> NodeBundle {
+    NodeBundle {
+        style: Style {
+            size: Size { width: Val::Percent(20.), height: Val::Percent(30.) },
+            align_items: AlignItems::Stretch,
```
Comment (Indy2222) [inline crates/controller/src/hud_components.rs:37]: Why are non-default values needed? At a first glance, it seems like only `size` is relevant in current state.

File: crates/controller/src/mouse.rs [hunk 3]
```diff
@@ -212,16 +212,47 @@ enum DragResolution {
...
+    hud: Query<(&GlobalTransform, &Node), With<HudVisibleNode>>,
+    mut mouse: ResMut<MousePosition>,
+) {
     let window = windows.get_primary().unwrap();
     mouse.set_position(
         window
             .cursor_position()
+            .and_then(|position| {
+                let win_size = Vec2::new(window.width(), window.height());
+                let found = hud.iter().find(|(transform, node)| {
```
Comment (Indy2222) [inline crates/controller/src/mouse.rs:226]: You can replace `.find()` with `.any()` which is what you want here.

File: crates/controller/src/mouse.rs [hunk 4]
```diff
@@ -212,16 +212,47 @@ enum DragResolution {
...
+                if found.is_none() { Some(position) } else { None }
+            })
             .map(|position| position / Vec2::new(window.width(), window.height()))
             .map(|normalised_position| normalised_position.clamp(Vec2::ZERO, Vec2::ONE)),
     );
 }
 
+fn is_in(
+    box_transform: &GlobalTransform,
+    box_size: &Node,
```
Comment (Indy2222) [inline crates/controller/src/mouse.rs:238]: `box_size` -> `node` would be a more appropriate argument name.

File: crates/controller/src/mouse.rs [hunk 5]
```diff
@@ -212,16 +212,47 @@ enum DragResolution {
...
+    win_size: Vec2,
+    mouse_position: Vec2,
+) -> bool {
+    // WARNING: This is because mouse y starts on bottom, GlobalTransform on top
+    let mouse_position = Vec2::new(mouse_position.x, win_size.y - mouse_position.y);
+
+    let box_size = box_size.size();
+    let box_transform: Vec3 = box_transform.translation();
+    // WARNING: This is because GlobalTransform is centered, width/2 to left and to right, same on vertical
+    let box_position = Vec2::new(box_transform.x, box_transform.y) - box_size / Vec2::splat(2.);
```
Comment (Indy2222) [inline crates/controller/src/mouse.rs:248]: * `Vec2::new(box_transform.x, box_transform.y)` -> `box_transform.xy()` should work.
* Isn't `box_size / Vec2::splat(2.)` the same as `box_size / 2.`?

File: crates/controller/src/mouse.rs [hunk 6]
```diff
@@ -212,16 +212,47 @@ enum DragResolution {
...
+
+    box_position.x <= mouse_position.x
```
Comment (Indy2222) [inline crates/controller/src/mouse.rs:250]: `mouse_position.cmpge(box_position) && mouse_position.cmple(box_position + box_size)` is more concise.

File: crates/controller/src/mouse.rs [hunk 7]
```diff
@@ -212,16 +212,47 @@ enum DragResolution {
...
             .map(|position| position / Vec2::new(window.width(), window.height()))
             .map(|normalised_position| normalised_position.clamp(Vec2::ZERO, Vec2::ONE)),
     );
 }
 
+fn is_in(
+    box_transform: &GlobalTransform,
+    box_size: &Node,
+    win_size: Vec2,
+    mouse_position: Vec2,
```
Comment (Indy2222) [inline crates/controller/src/mouse.rs:240]: This function does not care whether it is position of mouse or something else, so I would rename the argument to `point` or something generic.

File: crates/controller/src/mouse.rs [hunk 8]
```diff
@@ -212,16 +212,47 @@ enum DragResolution {
...
+                    is_in(transform, node, win_size, position)
+                });
+                if found.is_none() { Some(position) } else { None }
+            })
             .map(|position| position / Vec2::new(window.width(), window.height()))
             .map(|normalised_position| normalised_position.clamp(Vec2::ZERO, Vec2::ONE)),
     );
 }
 
+fn is_in(
```
Comment (Indy2222) [inline crates/controller/src/mouse.rs:236]: `is_in` -> `contains_point()` would IMHO be a better name.

File: crates/controller/src/mouse.rs [hunk 9]
```diff
@@ -212,16 +212,47 @@ enum DragResolution {
     Rect(ScreenRect),
 }
 
-fn update_position(windows: Res<Windows>, mut mouse: ResMut<MousePosition>) {
+fn update_position(
+    windows: Res<Windows>,
+    hud: Query<(&GlobalTransform, &Node), With<HudVisibleNode>>,
```
Comment (Indy2222) [inline crates/controller/src/mouse.rs:217]: I would experiment with making this a `SystemParam` to make the API nicer. The usage would look like this:

```rust
fn update_position(
    windows: Res<Windows>,
    hud: HudNodes,
    mut mouse: ResMut<MousePosition>,
) {
...
            .filter(|position| hud.contains_point(position))
}
```

File: crates/controller/src/mouse.rs [hunk 10]
```diff
@@ -212,16 +212,47 @@ enum DragResolution {
...
+    mut mouse: ResMut<MousePosition>,
+) {
     let window = windows.get_primary().unwrap();
     mouse.set_position(
         window
             .cursor_position()
+            .and_then(|position| {
```
Comment (Indy2222) [inline crates/controller/src/mouse.rs:224]: `.and_then()` is not needed, `.filter()` would suffice.

File: crates/controller/src/hud.rs [hunk 11]
```diff
@@ -79,3 +89,30 @@ fn process_events(
         }
     }
 }
+
+pub fn spawn_hud(
+    mut commands: Commands,
+) {
+    commands.spawn(
```
Comment (Indy2222) [inline crates/controller/src/hud.rs:96]: Is this needed? I think that setting absolute position to all three children and not using the parent would be more concise.

File: crates/controller/src/mouse.rs [hunk 12]
```diff
@@ -288,3 +319,6 @@ fn check_double_click(
         *last_click_position = Some(mouse_clicked.position());
     }
 }
+
+#[derive(Component)]
+pub(crate) struct HudVisibleNode;
```
Comment (Indy2222) [inline crates/controller/src/mouse.rs:324]: What about `HudTopVisibleNode`? Also, please document what this component means.

And I would place it alongside HUD as it is where it logically belongs.

**Review by Indy2222 — CHANGES_REQUESTED** [review]
I had a lot of comments, but the overall idea looks good!

File: crates/controller/src/mouse.rs (hunk 6 above)
Comment (Polostor) [reply crates/controller/src/mouse.rs]: It had to be using other method ( `some.cmpge(other).all()` ) but great advice!

File: crates/controller/src/mouse.rs (hunk 12 above)
Comment (Polostor) [reply crates/controller/src/mouse.rs]: It does belong to HUD, but as you written somewhere before you try to have all systems used in one plugin in single file.. which makes me confused as to where to put this `HudTopVisibleNode` and the newly created `SystemParam` called `HudNodes`..

Comment (Polostor) [reply crates/controller/src/mouse.rs]: And please where to document this? Is it sufficient in **code as a comment** eg. `\\\ Any HUD element that is the parent Node and should is not supposed to be draggable and other  interaction with in-game objects` or somewhere more detailed comment in documentation files like `CONTRIBUTING.md`?

Comment (Indy2222) [reply crates/controller/src/mouse.rs]: Code comment is sufficient. `Top-level non-transparent UI node. All such nodes are marked with this component and no descendants have it attached` or something like that.

Comment (Indy2222) [reply crates/controller/src/mouse.rs]: I would move this and the system parameter to the different module. The component does not have to be used from mouse whatsoever. It has to be temporarily public due to the limitations of system param derive, but that will change with Bevy v10.

**Review by Indy2222 — DISMISSED** [review]
SGTM. I will test it later today.

**Review by Indy2222 — APPROVED** [review]

**Pull request merged**